openedx/core/djangoapps/discussions/docs/decisions/0004-in-context-discussions-linking.rst

```diff
@@ -0,0 +1,131 @@
+need a way to:
+
+- Mark a Unit as discussable
+- Store configuration options for discussions
+- Automatically mark/unmark units as discussable when settings change
+- Associate a unit with a corresponding discussion id
+- Make this discussion ID available via an API
+
+
+Consideration
+-------------
+
+Most of these configuration entries would be right at home in the
+`DiscussionsConfiguration` model in `plugin_settings`, however since they need
+to be available during course import-export, they should be stored in the
+course object itself.
+
+There is already a way to associate a discussion ID with a Discussion XBlock
+using its usage key. This same mechanism can be used to associate a Unit usage
+key with a corresponding discussion id.
+
+However the current mechanism has a few issues. It is stored as a JSON
+structure in the `DiscussionsIdMapping` model which has course id and a mapping
+of the discussion id to the xblock usage key in a single dict.
+
+This is OK for the existing setup because this is just a caching mechanism and
+the source of truth for this mapping is the XBlock itself, which stores the
+discussion id. On course publish this information is cached to
+`DiscussionsIdMapping`.
+
+For the new discussions system though, this mapping would be the source of
+truth for the link between discussions and units, so we should use a model
+where each such link is encoded as a row in the database.
+
+Decision
+--------
+
+Since the discussions settings need to be stored in the course structure we
+should create a new JSON structure in the course that matches the structure
+of `plugin_settings`. This can then be used to store not just the settings
+for the inbuilt discussions provider, but for any discussions provider in the
+future.
+
+When a course is published, we can copy over all the `plugin_settings` to the
+course structure with the following structure:
+
+.. code-block:: JSON
+
+    {
+        "edx-next": {
+            "discussions_enable_in_context": bool,
+            "discussions_enable_graded_units": bool,
+            "discussions_group_at_subsection": bool,
+            "discussions_custom_visibility": bool,
+        }
+    }
+
+The `edx-next` key here represents the provider id, allowing for potentially
+multiple provider configs to coexist in case of switching providers etc.
+
+To store Unit-level discussions settings, we can simply add a boolean field
+to the Unit block that specifies whether it is discussable or not. To be
+consistent with the above names we can call this field `discussions_enabled`.
+
+When any of the above settings are changed, the code handling the API call will
+go through all the units and match them against the above criterion, i.e. if
+`discussions_enable_graded_units` is false, and an existing graded unit is
+discussable (whether because an existing unit was switched to gradable, or this
+setting was), then it's `discussions_enabled` flag will be turned off.
```

Comment:
Wait, does that mean changing a course setting could potentially trigger changing/publishing every Unit in the course? Won't that autopublish the contents of any Unit that has draft content in it as a side-effect?

Reply:
Hmm.. that is true, that would be a pretty bad side-effect. 

One alternative I see is to store the discussion enabled status in the database only. In this case it won't be imported/exported, but unless custom visibility is enabled, the import/export will anyway re-create these links. For custom visibility this data can be stored in some other course field, i.e. a list of keys with discussions enabled, which will need special handling on import-export.

Reply:
FWIW, I think it's fine to leave the field at the Unit level, as long as we don't do a mass reset when the config toggle changes at a high level. I don't think we should worry about keeping values in-sync, since people are going to have a chance to do weird things with OLX import anyway. The important thing is that the resulting configuration is _unambiguous_. So we can say, "Yeah, go ahead and flip this toggle at the Unit level, but we'll just ignore it unless `custom_visibility` is true." We can also reflect that in the Unit UX without having to do too much extra work.

Reply:
👍🏽 
Sure, I was hoping to just use that flag in the Unit as a single source of truth when determining if a unit has discussions enabled or not, but it seems doing that will just move complexity from one place to another.